contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,846 @@
+object's revision history. In practice, these PodTemplates will be filtered out
+prior to history maintenance.
+1. If the PodTemplate's ControllerRef matches the StatefulSet, the 
+StatefulSet controller will delete the PodTemplate.
+
+#### History Reconstruction
+In order to reconstruct the history of revisions to a StatefulSet, the 
+StatefulSet controller will do the following.
+
+1. If the StatefulSet's `.Spec.TemplateGeneration` is nil, the StatefulSet 
+has never been updated, and its history has never been initialized. This is 
+the state the object will be in when a cluster is first upgraded from a version
+that does not support StatefulSet update to a version that does. In this case,
+the controller will not enforce PodTemplate revisions. When creating Pods, 
+it will always use the StatefulSet's `.Spec.Template`. Otherwise, the controller 
+will continue as below.
+1. The controller will select all PodTemplates with a 
+`StatefulSetPodTemplateLabel` matching the `.Name` field of the StatefulSet.
+1. The controller will filter out all PodTemplates that do not contain a 
+ControllerRef matching the the StatefulSet. If the controller selects 
+PodTemplates that it does not own, it will report an error, but it will continue
+reconstructing the StatefulSet's history.
+1. The controller will filter out all PodTemplates that do not have a 
+`StatefulSetTemplateGenerationLabel` mapped to a valid revision. This can only 
+occur if the user purposefully deletes the label. In this case, the 
+controller will report an error, but it will continue reconstructing the 
+StatefulSet's revision history.
+1. For all the remaining PodTemplates, the controller will sort them in 
+ascending order by the value mapped to their `StatefulSetTemplateGenerationLabel`. 
+This will reconstruct a list of PodTemplates from oldest to newest. Note that, 
+as the revision is monotonically increasing for an individual StatefulSet, and
+as we use ControllerRef to mitigate selector overlap, the StatefulSet's history
+ is a strictly ordered set.
+
+#### History Maintenance 
+In order to prevent the revision history of the StatefulSet from exceeding 
+memory or storage limits, the StatefulSet controller will periodically prune 
+the oldest PodTemplates from the StatefulSet's revision history. 
+
+1. The StatefulSet controller will 
+[reconstruct the revision history](#history-reconstruction) 
+of the StatefulSet.
+1. The StatefulSet will remove any PodTemplates that correspond to created 
+Pods. There should be at most two of these, the PodTemplates corresponding 
+to the current and target revisions.
+1. If the number of PodTemplates in the StatefulSet's revision history is 
+greater than the StatefulSet's `.Spec.RevisionHistoryLimit`, the 
+StatefulSet controller will delete PodTemplates, starting with the head of 
+the revision history, until the size of the revision history is equal to 
+the StatefulSet's `.Spec.RevisionHistoryLimit`.
```

Comment:
In practice we do. There are at most two valid PodTemplates that can be applied. The current revision of the proposal indicates that we will not consider "any PodTemplated that correspond to created Pods"